### PR TITLE
Try to re-add stubs for ctypes

### PR DESCRIPTION
**Pull request by dgelessus**
Two years ago, some stubs for `ctypes` were added (#454) but removed soon afterwards, because they caused problems with correct code (#475). As I understand it, this was in part because of limitations in mypy at the time. Some of these limitations don't seem to exist anymore, so I tried to make the `ctypes` stubs work again.

I took the code in its last state before it was removed and tried to fix the issues reported in #475, as well as a few issues that I came across myself when using the stubs with my own code. Details on what exactly I've changed can be found in the individual commit messages. I'm sure there are still issues that I haven't noticed, so if anyone has a `ctypes`-using codebase and could try out these stubs, I would be happy to get some feedback.

Finally, I'd like to note that I don't have a lot of experience with mypy and writing `typing` stubs. Previously I've type-annotated some bits of my code and used typeshed through PyCharm's typechecker, but other than that this is my first time properly using mypy and `typing`. So if there's anything that I'm doing incorrectly or not according to convention, please let me know! 😃

### CODE REVIEW

File: stdlib/2and3/ctypes/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,257 @@
+# Stubs for ctypes
+
+from typing import (
+    Any, Callable, ClassVar, Iterable, List, Mapping, Optional, Sequence, Sized, Tuple, Type,
+    Generic, TypeVar, overload,
+)
+from typing import Union as UnionT
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: You should name this `_UnionT` so it doesn't get re-exported. (With the current definition, mypy would let users write `ctypes.UnionT`.)

File: stdlib/2and3/ctypes/__init__.pyi [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+RTLD_GLOBAL: int = ...
+RTLD_LOCAL: int = ...
+DEFAULT_MODE: int = ...
+
+
+class _DLL:
+    def __init__(self, name: str, mode: int = ..., handle: Optional[int] = ...,
+                 use_errno: bool = ..., use_last_error: bool = ...) -> None: ...
+    def __getattr__(self, name: str) -> Any: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Looks like there's also `__getitem__`.

File: stdlib/2and3/ctypes/__init__.pyi [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+class CDLL(_DLL): ...
+if sys.platform == 'win32':
+    class OleDLL(_DLL): ...
+    class WinDLL(_DLL): ...
+class PyDLL(_DLL):
+    _handle: int = ...
+    _name: str = ...
+    def __init__(self, name: str, mode: int = ...,
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Isn't the `__init__` just inherited from the base class? https://github.com/python/cpython/blob/2.7/Lib/ctypes/__init__.py#L389

File: stdlib/2and3/ctypes/__init__.pyi [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                 handle: Optional[int] = ...) -> None: ...
+
+class LibraryLoader(Generic[_DLLT]):
+    def __init__(self, dlltype: Type[_DLLT]) -> None: ...
+    def LoadLibrary(self, name: str) -> _DLLT: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Missing `__getattr__` and `__getitem__`.

File: stdlib/2and3/ctypes/__init__.pyi [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+def addressof(obj: _CData) -> int: ...
+def alignment(obj_or_type: UnionT[_CData, Type[_CData]]) -> int: ...
+def byref(obj: _CData, offset: int = ...) -> _cparam: ...
+_PT = TypeVar('_PT', bound=_PointerLike)
+def cast(obj: _CData, type: Type[_PT]) -> _PT: ...
+def create_string_buffer(init_or_size: UnionT[int, bytes],
+                         size: Optional[int] = ...) -> Array[c_char]: ...
+c_buffer = create_string_buffer
+def create_unicode_buffer(init_or_size: UnionT[int, str],
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Should be Text for Python 2. The actual code checks for (str, unicode).

File: stdlib/2and3/ctypes/__init__.pyi [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                 desc: Optional[str] = ...) -> OSError: ...
+def wstring_at(address: int, size: int = ...) -> str: ...
+
+class _SimpleCData(Generic[_T], _CData):
+    value: _T = ...
+    def __init__(self, value: _T = ...) -> None: ...
+
+class c_byte(_SimpleCData[int]): ...
+
+class c_char(_SimpleCData[bytes]): ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: The `__init__` argument here can also be an int.

File: stdlib/2and3/ctypes/__init__.pyi [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+class c_char_p(_PointerLike, _SimpleCData[Optional[bytes]]):
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Why is it Optional? If the value is inaccessible, it just segfaults (as I just found out experimentally).

File: stdlib/2and3/ctypes/__init__.pyi [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+class c_uint64(_SimpleCData[int]): ...
+
+class c_ulong(_SimpleCData[int]): ...
+class c_ulonglong(_SimpleCData[int]): ...
+
+class c_ushort(_SimpleCData[int]): ...
+
+class c_void_p(_PointerLike, _SimpleCData[Optional[int]]): ...
+
+class c_wchar(_SimpleCData[str]): ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Should this be Text?

File: stdlib/2and3/ctypes/__init__.pyi [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+class c_wchar_p(_PointerLike, _SimpleCData[Optional[str]]):
+    def __init__(self, value: UnionT[int, str] = ...) -> None: ...
+
+class c_bool(_SimpleCData[bool]):
+    def __init__(self, value: bool) -> None: ...
+
+if sys.platform == 'win32':
+    class HRESULT(_SimpleCData[Any]): ...  # TODO undocumented
+
+class py_object(Generic[_T], _SimpleCData[_T]): ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/ctypes/__init__.pyi]: Don't need the explicit Generic base class.

File: stdlib/2and3/ctypes/__init__.pyi (hunk 3 above)
Comment (dgelessus) [reply stdlib/2and3/ctypes/__init__.pyi]: Technically yes, but [the documentation](https://docs.python.org/2.7/library/ctypes.html#ctypes.PyDLL) lists the constructor as not accepting the `use_errno` and `use_last_error` kwargs. I don't know what the correct procedure is in cases like this - is the documentation or the code right?

File: stdlib/2and3/ctypes/__init__.pyi (hunk 7 above)
Comment (dgelessus) [reply stdlib/2and3/ctypes/__init__.pyi]: Not here - the `value` of a null pointer is `None`.

```python
Python 3.6.4 (v3.6.4:d48ecebad5, Dec 18 2017, 21:07:28) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
>>> import ctypes
>>> ccp = ctypes.c_char_p(None)
>>> ccp.value
>>> cwp = ctypes.c_wchar_p(None)
>>> cwp.value
```

**Comment by dgelessus** [comment]
OK, should all be fixed (along with a couple of other issues that I noticed/remembered).

File: stdlib/2and3/ctypes/__init__.pyi (hunk 3 above)
Comment (JelleZijlstra) [reply stdlib/2and3/ctypes/__init__.pyi]: I'd generally favor the code.

Comment (dgelessus) [reply stdlib/2and3/ctypes/__init__.pyi]: OK, I've removed the `__init__` from `PyDLL`, it should now inherit the one from `CDLL` again.

File: stdlib/2and3/ctypes/__init__.pyi [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    def __mul__(cls: _CDataMeta, other: int) -> Type[Array[_CT]]: ...
+    def __rmul__(cls: _CDataMeta, other: int) -> Type[Array[_CT]]: ...
+class _CData(metaclass=_CDataMeta):
+    _b_base: int = ...
+    _b_needsfree_: bool = ...
+    _objects: Optional[Mapping[Any, int]] = ...
+    @classmethod
+    def from_buffer(cls: Type[_CT], source: bytearray, offset: int = ...) -> _CT: ...
+    @classmethod
+    def from_buffer_copy(cls: Type[_CT], source: bytearray, offset: int = ...) -> _CT: ...
```
Comment (eric-wieser) [inline stdlib/2and3/ctypes/__init__.pyi]: Shouldn't this be `bytes` not `bytearray`?

Comment (dgelessus) [reply stdlib/2and3/ctypes/__init__.pyi]: Both of the `bytearray` types are inadequate actually - `from_buffer_copy` accepts any object that implements the [buffer protocol](https://docs.python.org/3/c-api/buffer.html), and similarly `from_buffer` accepts any writable buffer. `bytearray` and `bytes` are common examples of read-write and read-only buffer objects, but there are others like `memoryview` and `array.array`. All `ctypes` data objects (`_CData` subclasses) are writable buffers as well.

Does typeshed/`typing` have a type for "anything implementing the buffer protocol"? If not, we probably need to use `Union`s and list the most common stdlib buffer types.

Comment (ilevkivskyi) [reply stdlib/2and3/ctypes/__init__.pyi]: You can define a new protocol. I think there are already many examples in typeshed (grep for `Protocol`).

Comment (ilevkivskyi) [reply stdlib/2and3/ctypes/__init__.pyi]: Oh, it is C-level, in this case looks like a big union is the only option.

Comment (eric-wieser) [reply stdlib/2and3/ctypes/__init__.pyi]: Isn't `bytes` already treated by mypy as anything implementing the buffer protocol?

Comment (dgelessus) [reply stdlib/2and3/ctypes/__init__.pyi]: Fix here: #2610

**Pull request merged**